### ISSUE TITLE
MPEG-DASH - Intel Arc A380 - Youtube 4k60 HDR vp9 Videos fail to play - Kodi 20.2 and Git

### ISSUE THREAD

**Issue by ilikenwf** [xbmc/xbmc#23957]
### Describe the bug

My issue is more or less exactly https://github.com/xbmc/xbmc/issues/23649 but in regard to an Arc A380.

I just got an Arc A380, and I use Archlinux to run Kodi standalone using gbm. HDR should "just work" but it does not. For whatever reason, the video will load and then I either get a black screen or a ghost version of the kodi menu I was last on until I use the remote to stop the video or display the menu.

I get the audio, but no video...

## Expected Behavior
Playback working properly.

## Actual Behavior
See above

## Possible Fix
Disable HDR in inputstream adaptive settings - this allows proper playback.

### To Reproduce
Use VAAPI with an Arc A380

I've tried a couple different kernels, both stable and git versions of kodi and inputstream adaptive, and am at a loss. The h.264 version of the video plays fine, but does not allow for HDR from youtube.

Here's an example video that gives me trouble in webm/vp9 mode:

https://www.youtube.com/watch?v=VO8Bk03Xv90

### Debuglog
```
2023-10-18 03:44:49.468 T:8028    error <general>: GetPicture - avcodec_receive_frame returned failure
2023-10-18 03:44:49.468 T:8028    debug <general>: CVideoPlayerVideo - video decoder returned error
2023-10-18 03:44:49.529 T:8028    error <general>: ffmpeg[0x5577a5ae4540]: [vp9] Not all references are available
```

**Comment by ilikenwf** [comment]
I have altered the issue per the template, minus a few fields.

**Comment by ilikenwf** [comment]
> Thank you for using Kodi and our issue tracker. This is your friendly Kodi GitHub bot 😉
> 
> It seems that you have not followed the template we provide and require for all bug reports (or have opened a roadmap item by accident). Please understand that following the template is _mandatory_ and required for the team to be able handle the volume of open issues efficiently.
> 
> Please edit your issue message to follow our [template](https://github.com/xbmc/xbmc/blob/master/.github/ISSUE_TEMPLATE.md) and make sure to fill in all fields appropriately. The issue will be closed after one week has passed without satisfactory follow-up from your side.
> 
> This is an automatically generated message. If you believe it was sent in error, please say so and a team member will remove the "Ignored rules" label.

Please review, team member :)

**Comment by fritsch** [comment]
Where is your debuglog?

**Comment by smp79** [comment]
> 20.2 and Git

20.2 does not include the VP9 fix. What exactly "and Git" means?

**Comment by ilikenwf** [comment]
I just mean a build as recent as 671c9b94964

I'll try and get you a full log soon.

**Comment by smp79** [comment]
Is it playable in files mode with Kodi 21? I uploaded a part of that VP9 stream [here](https://www.dropbox.com/scl/fi/v1z2qtglylts1y9dhnvh0/videoplayback.mkv?rlkey=y21b4fg6ar182rwyu0k21e3m3&dl=1) It plays fine on my Intel iGPU.

**Comment by ilikenwf** [comment]
@fritsch see attached. Looks to me like either an ffmpeg issue or a DRM issue? @smp79 same behavior using your file.

Getting this over and over again:

```
2023-10-27 03:59:09.328 T:576     debug <general>: CDRMAtomicRequest::LogAtomicRequest - DRM Atomic Request:
2023-10-27 03:59:09.329 T:576     debug <general>: 
                                                   Object: plane	ID: 40
                                                     Property: SRC_X	ID: 9	Value: 0
                                                     Property: SRC_Y	ID: 10	Value: 0
                                                     Property: SRC_W	ID: 11	Value: 251658240
                                                     Property: SRC_H	ID: 12	Value: 141557760
                                                     Property: CRTC_X	ID: 13	Value: 0
                                                     Property: CRTC_Y	ID: 14	Value: 0
                                                     Property: CRTC_W	ID: 15	Value: 3840
                                                     Property: CRTC_H	ID: 16	Value: 2160
                                                     Property: FB_ID	ID: 17	Value: 283
                                                     Property: CRTC_ID	ID: 20	Value: 80
2023-10-27 03:59:09.348 T:1759    debug <general>: CVideoPlayerVideo::CalcDropRequirement - hurry: 0
2023-10-27 03:59:09.348 T:1759    error <general>: ffmpeg[0x561977149b80]: [vp9] Not all references are available
```

[kodi.log.zip](https://github.com/xbmc/xbmc/files/13187045/kodi.log.zip)

**Comment by ilikenwf** [comment]
Oh please revisit this my friends. HDR doesn't work on the aforementioned video of Obi-Wan vs Darth Vader, and many others, nor on the example posted by @smp79, but for some weird reason this one does???

https://www.youtube.com/watch?v=Hu_iaEbyUmg

**Comment by ilikenwf** [comment]
Further...if I disable VP9 VAAPI, I can play HDR videos.

**Comment by fritsch** [comment]
File a bug with intel. We only use VAAPI as designed. If the ARC and its driver behave differently -> needs fixing on that level. Here: https://github.com/intel/libva

**Comment by ilikenwf** [comment]
Are you sure it's not an issue with ffmpeg instead? In addition, historically, things like this do happen - https://github.com/xbmc/xbmc/issues/23649

**Comment by ilikenwf** [comment]
I'm testing again with the git version of inputstreamadaptive and the 21.0-BETA1 d92683f457d

**Comment by ilikenwf** [comment]
No luck sadly.

**Comment by ilikenwf** [comment]
So this is interesting...if I fetch it with yt-dlp, the webm will then play back. Would that indicate something with inputstream adaptive, or is Google breaking vp9?

**Comment by ilikenwf** [comment]
Could this be some kind of DRM issue?

**Comment by fritsch** [comment]
As pointed out by others, it even does not work in file mode for you, while it works for other users running the same ffmpeg, but a different driver for vaapi on different hardware. So from that: not much left.

So, I see absolutely no inputstream issue as a simple file produces the same issue for you on your hw combination. Not sure why you add more complexity and don't just focus on the file playback issue?

**Comment by atoulmin** [comment]
4K60 HDR has never worked in Intel with Kodi.

**Comment by fritsch** [comment]
https://github.com/xbmc/xbmc/issues/23957#issuecomment-1769768523 your comment is not helpful at all when we talk about above sample.

**Comment by ilikenwf** [comment]
> 4K60 HDR has never worked in Intel with Kodi.

Then why, when using GBM on Wayland, does it work inside a webm container when I grab it from youtube, but not when I streaim it directly/with inputstream adaptive?

**Comment by fritsch** [comment]
Ignore that comment - just a usual bug high jacker ...

**Comment by ilikenwf** [comment]
I've not stopped thinking about this, but I haven't had a chance to evaluate the approach. I guess get a definitive log of both with and without HDR, and then post it where - here, or to one of the ffmpeg repos, or to intel-media-driver? Suggestions welcome.

**Comment by ilikenwf** [comment]
So using slyguy's youtube addon, I can play 4k HDR without issues...using the "future" build of the official youtube addon, I have to disable HDR in the feature selection area for mpeg-dash to be able to play anything with HDR as an option, or else I get my blank screen issue...

I still feel like this has to either be:

* a stream selection issue in the youtube plugin
* or some issue with the intel media driver...(i run the git version)

**Comment by ilikenwf** [comment]
I figured it out!

I  had whitelist resolutions enabled and somehow that was causing it...of course now I have to figure out how best to get Kodi to upscale it to full screen size when there's black bars all around.

**Comment by ilikenwf** [comment]
...that said scaling shouldn't cause HDR to break? Likewise the video seems kinda laggy.

**Comment by ilikenwf** [comment]
Allowing pulldown and pull up allows me to keep using the whitelist...

I  just wish that using the display syncing feature didn't cause me to sometimes lose a few seconds of audio/video before the tv has changed to another mode...I use this option because when I run in 4160 4k mode, I have black bars on everything, but don't want to run at 3840x2160 by default because then any 16:10 videos I watch won't display properly...thoughts?

**Comment by ilikenwf** [comment]
For whatever reason this is cropping up again...even with whitelist disabled. The yt-dlp fetched videos play fine and are HDR10, while the streamed version through inputstream.adaptive fails.

I've tried this both with and without the "Use Display HDR Capabilities" checked.

Disabling VAAPI VP9 hardware decoding does fix it, but this defeats the whole point of having an intel Arc Card.

**Comment by ilikenwf** [comment]
The issue is somewhat inconsistent at times as after playing the downloaded file and re-enabling HDR in Youtube, it worked...strange. All I did was enable Automatic + Quality Selection instead of Automatic in the beta version of the youtube plugin.

**Comment by ilikenwf** [comment]
This whole thing is very strange. I've noted that one specific file I test with will enable HDR if I use a nightly snapshot of Libreelec, none of the others do...using good cables etc...

Meanwhile if I use a nightly build of kodi with the latest versions of the intel drivers and such, still no HDR...it makes no sense.

I like the Arc card for a lot of reasons but I'm tempted to throw my 1080ti back in and use nvdec for HDR10 playback.

**Comment by ilikenwf** [comment]
```
2023-12-15 23:54:17.730 T:503      info <general>: [display-info] make: 'Sony' model: 'SONY TV  *30'
2023-12-15 23:54:17.730 T:503      info <general>: [display-info] supports hdr static metadata type1: true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info] supported eotf:
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   traditional sdr: true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   traditional hdr: false
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   pq:              true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   hlg:             true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info] luma min: '0.026991393' avg: '2166.8088' max: '2166.8088'
2023-12-15 23:54:17.730 T:503      info <general>: [display-info] supported colorimetry:
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   xvycc_601:   true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   xvycc_709:   true 
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   sycc_601:    true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   opycc_601:   true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   oprgb:       true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   bt2020_cycc: false
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   bt2020_ycc:  true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   bt2020_rgb:  true
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   st2113_rgb:  false
2023-12-15 23:54:17.730 T:503      info <general>: [display-info]   ictcp:       false
```

**Comment by ilikenwf** [comment]
So youtube HDR playback is still fully broken, but I found arch is packaging kodi with the gl instead of gles backend.

**Comment by ilikenwf** [comment]
So 100% this is an issue with VAAPI vp9 hardware decoding. If I turn it off and use software, HDR playback works, as does just the general video playback on youtube...

Oddly one of my test videos that's in HDR, downloaded from youtube, still doesn't switch to HDR mode so that's an odd outlier.

**Comment by fritsch** [comment]
Last part is wrong. Kodi decodes and imports 10 bit precision into OpenGL, even uses 16 bit precission textures internally. Final visuals depend on the windowing system, but has zero to do with your issue here.

For the rest I can't follow - from "it works" to "scaling should not be an issue", "whitelisting fixed" it ... sorry.

**Comment by ilikenwf** [comment]
Ok, so forget everything I said until my messages for today, sorry, I haven't slept in like 23 hours...new TV gave me the motivation, haha.

So, I'm 99% working with HDR playback, the primary issue was that kodi on my distro is built using gl instead of gles.

Past that, vp9 hardware accelerated playback seems to be the cause of the issue with HDR playback of youtube videos. When disabled in favor of software rendering, it usually works. Too bad since my Arc card supports hardware decode of vp9.

Glad to hear the last part of my previously, poorly written comment was wrong. So far everything looks great!

I was saying I would like to assist in whatever way possible to fix the VP9 hardware decoding issue on my Arc card, if anyone thinks it worthwhile.

**Comment by fritsch** [comment]
As far as I know, no one has this issue with VAAPI without ARC. That's why I asked you some months back to take the bugreport to intel's vaapi as they for sure don't want "specific" ARC behaviour in ffmpeg ... which drives the VAAPI way of doing things entirely nuts. 

Same in kodi btw. Fully standard VAAPI decoding, combined with an opensource render path via mesa ... if anything there is special for ARC, it needs fixing and not workarounding.

**Comment by ilikenwf** [comment]
I appreciate that - at the time I had 0 HDR output whatsoever though, so I was focused more on that, this situation I have now is much more tolerable at least..

I'll try to report it to Intel and see if I can get any traction. Was this libva or was it the intel-media-driver repo?

**Comment by fritsch** [comment]
It's the driver's repo. libva is just the VAPI, while the driver implements the bits. It would be good if you could provide an offline sample. Best shared with a proper log of a non-arc vaapi experience compared with yours. Every layer in between (be it inputstream, etc.) should be taken out of the chain to not cause confusion for upstream.

If ffmpeg's implementation is at failure here or - which is even llikely - that something could be made "more safe" in terms of references, it is always good to have the lowest burdon to reproduce it.

Throwing kodi + inputstream + whatever on them won't lead to a fast solution.

**Comment by ilikenwf** [comment]
The interesting part with the offline sample is that I can't reproduce the "black screen" playback issue unless I'm streaming it, downloading the VP9 video will usually result in either video occasionally freezing up, or just normal playback without HDR mode kicking in...fun stuff!

I don't think it's Kodi's fault, but I do know that the community around Kodi is a lot more friendly and well informed on issues such as these :)

**Comment by fritsch** [comment]
Can you reproduce the blackscreen with let's say mpv? Which is using vaapi as well?

**Comment by ilikenwf** [comment]
I will have to give that a shot! I did have this machine setup to just run Kodi + GBM but did install X and a simple DE last week for other debugging adventures.

**Comment by fritsch** [comment]
In general there are some things to distinguish:
Kodi doing decoding (vaapi) (1)
Kodi importing the surfaces(mesa) (2)
Kodi switching to "HDR" mode (GBM only). (3)

I am always a bit confused  reading your repots ... now I have the feeling the issue is with the latter part (3). If that's the case - please don't file a bug with intel people that would not be their beef.

**Comment by ilikenwf** [comment]
I'll do further testing as time permits, apologies for being all over the place with my descriptions. 

It could be specific to HDR or it may be an issue with VP9, really on any of those layers.

**Comment by fritsch** [comment]
Take your time. It's the same for me - due to my everyday job I have a very short focussed attention time on things, which makes it then even more easy to confuse me ... Good Luck :-).

**Comment by ilikenwf** [comment]
I've also got a pull request in to the Archlinux package so they can support HDR...their wiki implies it can but without gles it cannot...fun stuff.

**Comment by fritsch** [comment]
HDR needs a careful defition. Do you mean:
Decoding 10 bit content including metadata?
Do you mean rendering the said content with e.g. tonemapping?
Do you mean that your TV "shows" HDR mode?

Especially the latter - take a great grain of salt ... if you look into the details, e.g. what bit depth are the visuals, what additional patches are needed to get v4l2 on intel running. what are the difference to "normal" vaapi + GBM, etc. ... not really much is left ;-)

So - therefore if you find something you like, keep it, don't talk too much with people knowing the "swap chain" if you really get the details on screen.

**Comment by ilikenwf** [comment]
I mean decoding 10 bit content including metadata and pushing it out to my brand new 2023 bravia. With gles and this PR https://github.com/xbmc/xbmc/pull/24160, and vp9 hardware decoding disabled for the arc, everything for the most part "just works" including switching the display into HDR mode with no tonemapping done by kodi.


I've tested with various real test files, including a sony mp4, downloaded youtube videos that ffmpeg confirms are 10 bit hdr, streaming those same videos, and also with "other" videos "found" through privateers. So far I've tested vp9, mp4, and hevc, in various containers.

**Comment by fritsch** [comment]
Can we focus on this file: https://github.com/xbmc/xbmc/issues/23957#issuecomment-1769768523 please?

I looked at your log again and can see that the hwdecoder errors out in its endframe method. So cause of SW-Decoding + GLES is perfectly fine, we can basically rule out anything on the final render part and as we see decoding errors, we are back at vaapi.

So - now - don't jump anywhere else -> focus on this mkv file. Does it decode properly with mpv or better with ffmpeg plain using vaapi? I still suspect an issue in your vaapi driver.

**Comment by fritsch** [comment]
I want that output (don't forget the final hyphen), you need an ffmpeg with  vaapi compiled in:

ffmpeg -hwaccel vaapi -loglevel verbose -i videoplayback.mkv -f null -

If this also fails, this output can be directly file with ffmpeg.

**Comment by fritsch** [comment]
If the above yields an issue, replace verbose with debug - and file a bug with ffmpeg. if it works flawlessly, try:

```
diff --git a/xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp b/xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp
index fb7606e0d0..9aca123371 100644
--- a/xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp
+++ b/xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp
@@ -712,7 +712,7 @@ bool CDecoder::Open(AVCodecContext* avctx, AVCodecContext* mainctx, const enum A
   else if (avctx->codec_id == AV_CODEC_ID_HEVC)
     m_vaapiConfig.maxReferences = 16;
   else if (avctx->codec_id == AV_CODEC_ID_VP9)
-    m_vaapiConfig.maxReferences = 8;
+    m_vaapiConfig.maxReferences = 16;
   else if (avctx->codec_id == AV_CODEC_ID_AV1)
     m_vaapiConfig.maxReferences = 21;
   else
   ```
ffmpeg uses 16 surfaces for decoding - we only 8.

**Comment by ilikenwf** [comment]
The issue with all that being that videoplayback.mkv from the previous comment has always played properly both with and without vaapi. The errors you saw in previous logs were either from a playback of a youtube stream from inputstream adaptive or are becasue the videoplayback.mkv file is truncated and not complete in terms of the video it is sourced from, perhaps?

It's only when streaming direct from youtube through inputstream.adaptive that I run into this issue, files on disk do not seem to have this problem. 

I don't have your output yet from mpv - I can still get it if I need to. Does mpv allow for streaming direct from youtube? If not I'm not sure it would be a valid test since this issue only happens when streaming.

It also appears AV1 is a problem with VAAPI for me...This video crashes kodi entirely, I'll share the log below...I have a yt-dlp version of it on disk, and it properly plays with and without hardware decoding, with HDR working properly as well. With AV1 hardware decoding disabled, it plays perfectly.

https://www.youtube.com/watch?v=0nTO4zSEpOs

https://gist.github.com/ilikenwf/c0e2d35f7a015f504d7e88790a5e59a3

I can test with mpv for you but since hardware vp9 and av1 decoding work on downloaded youtube videos and not on those coming from inputstream adaptive it leads me back to thinking it's probably something with inputstream.adaptive or youtube even though the people involved in those think otherwise.

**Comment by ilikenwf** [comment]
It appears the crash only happens if I allow refresh rate adjustment on start/stop...so could it be that and video geometry causing the trouble?

I'm starting to agree that it feels like it has to be either ffmpeg or vaapi, or the intel driver itself.

**Comment by fritsch** [comment]
Sorry, but you jump again into a different codec, into a different topic, furthermore  your logfiles are not of any use, if it was your first day reporting kodi issue okay .... but yeah this here will lead to nowhwere, sorry:
2023-12-17 16:19:01.320 T:553434    info <general>: Disabled debug logging due to GUI setting. Level 0.


And while you bring in yet another thing with refreshrates,  your log indicates that 
2023-12-17 16:19:17.171 T:553505    info <general>: Display resolution ADJUST : 3840x2160 @ 59.940063 Hz (23) (weight: 0.000)

is used as refreshrate is a bit beyond me. But VAAPI was not in use at all here. Due to your insufficient log ... most likely nobody can help you. I see additionally that you use an insanely large cache / memory configuration - if you have additionally massive network issues, that even opens a third path of issues.

So in short - with the provided logs, with the changing theories and even codecs ... I can't help you further.

**Comment by ilikenwf** [comment]
I will test the various layers, I will enable debug logging, etc. The cache thing is an old config I need to remove from a time when I had lower bandwidth with the ISP.

I'll reconfigure my whitelist as well. I'll test VP9 and keep that constrained to this issue, and move on to AV1 when time permits.

**Comment by ilikenwf** [comment]
I will grab the av1 and vp9 test clips from here and test them with hardware decoding. If those work, and youtube videos via stream fail, I'll go from there:

https://kodi.wiki/view/Samples#4K_(UltraHD)_Formats

**Comment by ilikenwf** [comment]
@fritsch I apologize for tagging you, and also frustrating you. But I really have trouble believing this is an ffmpeg or driver layer issue at this point, or even a vaapi issue. I think it is something to do with inputstream.adaptive.

Though it works for playback with vaapi rendering turned off, when turned on HDR fails. I'm thinking there may be some kind of metadata missing or set to invalid values, perhaps?

If I use yt-dlp to just get the direct url of an HDR youtube video, the video stream (minus audio since not using mpd) plays perfectly and enables HDR mode.

If done through youtube, and thus inputstream.adaptive, I encounter the playback problem. This appears to happen with both AV1 and VP9.

I previously had an issue open on the inputstream.adaptive tracker but was dismissed.

I know I'm driving everyone crazy but I do believe something either in youtube or inputstream.adaptive is the culprit.

**Comment by fritsch** [comment]
Things are easy: Less text - more complete debuglogs.

**Comment by ilikenwf** [comment]
Fine, then. Here's my control test.

I am testing using the YouTube Video https://www.youtube.com/watch?v=jECIZpEsX5s

I also am testing using the yt-dlp version of it here:

The complete log below is from playback of the youtube streaming version, 4k HDR. The downloaded version plays as expected. The streamed version results in a blank screen, eith glitches and artifacts of the kodi menu displaying until I stop it. 

You may want to grep for "ffmpeg" as I turned on most of the component logging when I did this to be complete. The log is a bit over 5MB compressed:


[kodi.log.zip](https://github.com/xbmc/xbmc/files/13701681/kodi.log.zip)

The common theme is: 
ffmpeg[0x559ed94605b0]: [vp9] Not all references are available

**Comment by ilikenwf** [comment]
@fritsch mpv under plasma + wayland, while not currently passing through HDR unfortunately, has no problem playing back the video from youtube directly. I'm leaning toward it being more an issue with the stream itself after going through inputstream.direct...

Unfortunately I am unaware of many ways to test this directly unless I maybe using the slyguy youtube fork since it uses yt-dlp.

**Comment by ilikenwf** [comment]
Meanwhile, if I use my little addon that grabs all possible stream urls from a page, and pass it any HDR youtube URI, the video plays perfectly, albeit without audio currently because I don't build an HLS stream.

I really think this is some metadata or issue with how streams get handled by either the youtube plugin or inputstream.adaptive.

**Comment by ilikenwf** [comment]
@fritsch I opened an issue with the intel media driver team, after looking at one of their previous similar bugs I wonder if this isn't an issue with vp9 decoding, either with intel or ffmpeg, since vp9 apparently can have different geometry per frame...it could explain the log I posted in the bug (in which I mentioned this thread).

**Comment by ilikenwf** [comment]
I still question if it isn't an issue with inputstream.adaptive, if not ffmpeg or the media-driver...hopefully we will find out.

**Comment by ilikenwf** [comment]
@fritsch 

An interesting curiosity is that even if I turn off HDR display mode switching, that with HDR video allowed, and attempting to playback a youtube HDR video with hardware acceleration enabled for vp9, that the video still fails to play and just glitches the display. 

I'm not sure if this might be a core issue, a driver issue, or perhaps something to do with incomplete or extraneous HDR metadata?

Again, if I play the video directly using yt-dlp to grab the HDR video stream URI and play just that, it does work, even streaming from youtube...

I suppose it could also be a stream selection issue with youtube or inputstream.adaptive?

**Comment by rsrmean** [comment]
Hello,

I'd like to chime in here. I have the same issues while trying to play from Youtube with MPEG-DASH enabled.

**Hardware**: Intel i5-1135g7 w/ integrated Intel Iris Xe Graphics

VP9-Videos from the Youtube-Addon can be played with VA-API enabled as long as they aren't HDR. When they are HDR, only Audio is played. The menu glitches. Stopping Playback recovers the menu and Kodi can be used normally. HDR Videos can be played when disabling VA-API, but my little computer struggles with SW Decode.

I used yt-dlp to extract the video stream URL. When I send the video stream URL directly to Kodi, playback works without any issues with VAAPI enabled (except the missing sound ;)). The TV also indicates that it switched to "HDR-Mode".


**Tested Video:** [https://www.youtube.com/watch?v=mkggXE5e2yk](url)
**Debug Log for broken playback**: [vp9_broken.log](https://github.com/user-attachments/files/17250625/vp9_broken.log)


**Direct Video Stream URL:** [https://rr1---sn-oxujvavbox-jbod.googlevideo.com/videoplayback?expire=1728012579&ei=wwz_Zs3kI_SL6dsP3I7o4Aw&ip=2a0a%3Aa543%3A5503%3Af1%3A1c1d%3Ae1db%3A9b0f%3A6ce5&id=o-ABqCs0PU6pvGWOHdfHKBVlM_RDXRfGoUQIpzzwS01Z-j&itag=337&source=youtube&requiressl=yes&xpc=EgVo2aDSNQ%3D%3D&mh=br&mm=31%2C29&mn=sn-oxujvavbox-jbod%2Csn-4g5edn6r&ms=au%2Crdu&mv=u&mvi=1&pcm2cms=yes&pl=48&vprv=1&svpuc=1&mime=video%2Fwebm&rqh=1&gir=yes&clen=556965136&dur=158.591&lmt=1726285201837575&mt=1727989739&fvip=5&keepalive=yes&fexp=51300761&c=IOS&txp=4502434&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cxpc%2Cvprv%2Csvpuc%2Cmime%2Crqh%2Cgir%2Cclen%2Cdur%2Clmt&sig=AJfQdSswRQIgI6AcMvHhcAu2Qdz2yim10G_RiM8RZvZMPcwdftxWlGsCIQD8Ewu9ZPajljNC2vIAnWhlGAd1GmyP-w9aNKWNE_N9Jg%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl&lsig=ACJ0pHgwRQIgL8E6dMeBgj6lf_sbV8qlZha4GeAiaGiwGtuY9PPZyjsCIQC0dtnq0Yk-25VISLrEQY9osQJLcf1-2ROlWk1cA0bgDA%3D%3D](url)
**Debug Log for working playback**: [vp9_ok.log](https://github.com/user-attachments/files/17250628/vp9_ok.log)

@fritsch I hope you are still watching this. If there is anything more needed, please let me know!

**Comment by neo1973** [comment]
The linked video plays without issues using the YouTube plugin with HDR on an AMD Ryzen 7 5700U. VAAPI is enabled. No menu glitching. Therefor probably an Intel driver issue.

**Comment by ilikenwf** [comment]
If I recall I opened a ticket and they wanted more info that I couldn't provide at the time, and I haven't looked at it in a while.

https://github.com/intel/media-driver/issues/1758

**Comment by MoojMidge** [comment]
> Therefor probably an Intel driver issue.

Maybe, but there is this in the non-working log:

```
2024-10-03 23:26:34.299 T:870795   error <general>: ffmpeg[0x5e8f1c8d9d40]: [vp9] Failed to end picture decode issue: 18 (invalid parameter).
2024-10-03 23:26:34.299 T:870795   error <general>: ffmpeg[0x5e8f1c8d9d40]: [vp9] Not all references are available
2024-10-03 23:26:34.300 T:584893    info <general>: Skipped 85 duplicate messages..
```

Perhaps the right VP9 profile is not being passed through correctly?

Could try loading the MPEG-DASH manifest file in mpv and see how it handles it compared to Kodi.

**Comment by rsrmean** [comment]
Unfortunately, I'm stuck here with mpv:

`[ffmpeg/demuxer] dash: Filename extension of '.kodi/temp/plugin.video.youtube/mkggXE5e2yk.mpd' is not a common multimedia extension, blocked for security reasons.`
`[ffmpeg/demuxer] If you wish to override this adjust allowed_extensions, you can set it to 'ALL' to allow all`

And I can't find a way to let mpv pass that option to ffmpeg...

**Comment by MoojMidge** [comment]
Will need to load it from a http/https source. Otherwise can try renaming to `mkggXE5e2yk.xml` or setting up mpv as an external player in Kodi and loading it directly from the plugin

**Comment by rsrmean** [comment]
That command line option didn't work, but playing from a https source did. It plays fine - although I'm not sure if it plays in HDR mode. The TV didn't indicate being switched to HDR.

`AO: [alsa] 44100Hz stereo 2ch s32`
`VO: [gpu-next] 3840x2160 vaapi[p010]`

[mpv.log](https://github.com/user-attachments/files/17257562/mpv.log)

**Comment by ilikenwf** [comment]
To run in HDR you have to use gbm under Linux, if I'm not mistaken, so HDR may not be possible to test with mpv unless something has changed.

**Comment by ilikenwf** [comment]
...and running a git build of kodi 21.5.4 with the intel media drivers, gmmlib, etc all git versions, and the git version of youtube's addon, kodi now just outright crashes with HDR video playback from youtube unless I disable av1 and vp9 hardware acceleration.

**Comment by ilikenwf** [comment]
So I'm running kodi 21.5.4 as I said, using the ffmpeg 7.1 patch as well as the hdr metadata patch that for some reason never got merged in, but does in fact fix HDR for non-youtube videos (local at least).

Attached is a crashlog from my attempt to play: https://www.youtube.com/watch?v=njX2bu-_Vw4

[kodi_crashlog-20241014_032418.log.zip](https://github.com/user-attachments/files/17361034/kodi_crashlog-20241014_032418.log.zip)

**Comment by jwmay2012** [comment]
I have this exact issue on multiple intel devices. 
VP9 HDR content with HW decoding from youtube plays audio only, no video. SW decoding works with HDR as indicated on the TV, but stutters on slow devices. 
If i download the 4k 60fps HDR VP9 video with `yt-dlp` and play the file directly, HW decoding works fine.

- 12th gen N97
- 13th gen i5-13500H
- 14th gen Core 3 100U

I was attempting to upgrade from an Odroid N2+ which could handle the youtube addon streamed VP9 HDR just fine.

**Comment by ilikenwf** [comment]
This really makes no sense and we can't get anyone to pay any attention to it. I just tested with the full gl instead of gles build since the current master branch now includes a patch that enables HDR on gl, but still, no dice - black screen with acceleration enabled. So frustrating!

@jwmay2012 @rsrmean you had any luck lately?

**Comment by rsrmean** [comment]
No. I'm on Debian with an old driver, so I'll check again when Trixie is released.

I can download HDR Videos from Youtube on my phone and then send them to Kodi, works flawlessly with VA-API enabled.

**Comment by ilikenwf** [comment]
@fritsch @rsrmean 

**Update with new findings and a fresh debug log:**

While using hardware decoding, I tested playback on the current beta version of the YouTube addon.

First, I tried this AV1 HDR video:  
https://www.youtube.com/watch?v=njX2bu-_Vw4  
Playback triggered a segmentation fault (crash), and I've attached the corresponding debug log.

Next, I played this VP9 video:  
https://www.youtube.com/watch?v=jECIZpEsX5s  
As usual, the video didn't display—no crash this time. If I manually switch to the non-HDR stream via the video settings, playback works correctly.

Disabling HDR entirely in the plugin settings also resolves the issue—everything plays back as expected.

**Notes:**  
- The attached debug log has full verbosity enabled.  
- One odd thing I noticed: `inputstream.adaptive` appears to be binding to the machine’s LAN IP instead of `localhost`. Any reason why that might be?

**This stands out to me:**
`Failed to end picture decode issue: 18 (invalid parameter).`

[kodi_crashlog-20250621_231117.log.zip](https://github.com/user-attachments/files/20851069/kodi_crashlog-20250621_231117.log.zip)

**Comment by rsrmean** [comment]
> No. I'm on Debian with an old driver, so I'll check again when Trixie is released.
> 
> I can download HDR Videos from Youtube on my phone and then send them to Kodi, works flawlessly with VA-API enabled.

Checked it, still doesn't work. I'll resort to downloading files and sending them to Kodi, which still works perfectly fine.
